File: day2/objc/day2-part1.m
#import <Foundation/Foundation.h>

NSInteger safe(NSInteger value) {
    return value + 1;
}

NSInteger unsafe() {
    return 0;
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *filePath = @"../input.txt";
        NSString *fileContents = [NSString stringWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:nil];
        
        NSArray *array = [fileContents componentsSeparatedByString:@"\n"];
        NSMutableArray *reports = [NSMutableArray array];
        
        for (NSString *element in array) {
            NSArray *someArray = [element componentsSeparatedByString:@" "];
            [reports addObject:someArray];
        }
        
        BOOL (^differenceWithinRange)(NSInteger, NSInteger) = ^BOOL(NSInteger left, NSInteger right) {
            NSInteger difference = labs(left - right);
            return difference >= 1 && difference <= 3;
        };
        
        BOOL (^doesNotSatisfyRules)(NSInteger, NSInteger, BOOL) = ^BOOL(NSInteger left, NSInteger right, BOOL isAscending) {
            if (isAscending) {
                return left > right || !differenceWithinRange(left, right);
            } else {
                return left < right || !differenceWithinRange(left, right);
            }
        };
        
        NSInteger totalSafeReports = 0;
        for (NSArray *report in reports) {
            BOOL isAscending = NO;
            
            NSInteger firstLevel = [report[0] integerValue];
            NSInteger secondLevel = [report[1] integerValue];
            
            if (firstLevel == secondLevel) {
                continue;
            } else if (firstLevel < secondLevel) {
                isAscending = YES;
            }
            
            if (!differenceWithinRange(firstLevel, secondLevel)) {
                continue;
            }
            BOOL shouldSkipOuterLoop = NO;
            for (int i = 1; i < ([report count] - 1); i++) {
                NSInteger leftLevel = [report[i] integerValue];
                NSInteger rightLevel = [report[i+1] integerValue];

                if (doesNotSatisfyRules(leftLevel, rightLevel, isAscending)) {
                    shouldSkipOuterLoop = YES;
                    continue;
                }
            }
            
            if (shouldSkipOuterLoop) {
                continue;
            }
            
            totalSafeReports += 1;
        }
        
        NSLog(@"%lu", totalSafeReports);
    }
    
    return 0;
}

File: day2/swift/day2-part1.swift
import Foundation

let directoryPath = FileManager.default.currentDirectoryPath

// Go back one directory (use ".." to navigate up)
let parentDirectoryPath = (directoryPath as NSString).deletingLastPathComponent  // Go one directory up
let filePath = parentDirectoryPath + "/input.txt"

let fileContents = try! String(contentsOfFile: filePath, encoding: .utf8).components(separatedBy: "\n")

let reports = fileContents.map { $0.split(separator: " ").compactMap { Int($0)} }
let differenceWithinRange: (Int, Int) -> Bool = { left, right in
    let difference = abs(left - right)
    
    return (1...3).contains(difference)
}

let doesNotSatisfyRules: (Int, Int, Bool) -> Bool = { left, right, isAscending in
    if isAscending {
        return left > right || !differenceWithinRange(left, right)
    } else {
        return left < right || !differenceWithinRange(left, right)
    }
}

print(reports.reduce(0) { x, report in
    var isAscending = false
    
    let firstLevel = report[0]
    let secondLevel = report[1]
    
    if firstLevel == secondLevel {
        return unsafe()
    } else if firstLevel < secondLevel {
        isAscending = true
    }
    guard differenceWithinRange(firstLevel, secondLevel) else { return unsafe() }
    
    for index in 1 ..< report.count - 1 {
        let leftLevel = report[index]
        let rightLevel = report[index + 1]
        
        if doesNotSatisfyRules(leftLevel, rightLevel, isAscending) { return unsafe() }
    }
    
    return safe(acc: x)
})

private func safe(acc value: Int) -> Int {
    value + 1
}

private func unsafe() -> Int {
    0
}
